oprf/group/group.go

```diff
@@ -0,0 +1,344 @@
+	}
+
+	hasher, err := suite.Get([]byte(c.dst))
+	if err != nil {
+		return nil, err
+	}
+
+	return eccHasher{c, hasher}, nil
+}
+
+// HashToPoint produces a new point by hashing the input message.
+type HashToPoint interface {
+	Hash(msg []byte) (*Point, error)
+}
+
+type eccHasher struct {
+	suite  *Ciphersuite
+	hasher h2c.HashToPoint
+}
+
+func (h eccHasher) Hash(in []byte) (*Point, error) {
+	q := h.hasher.Hash(in)
+	x := q.X().Polynomial()
+	y := q.Y().Polynomial()
+
+	p := &Point{h.suite, new(big.Int), new(big.Int)}
+	p.x.Set(x[0])
+	p.y.Set(y[0])
+
+	// TODO: failing here
+	//if !p.IsValid() {
+	//	return nil, errors.New("invalid point")
+	//}
+
+	return p, nil
+}
+
+// HashToGroup performs a transformation to encode bytes as a Element object in the
+// group.
+func (c *Ciphersuite) HashToGroup(in []byte) (*Point, error) {
+	hasher, err := getH2CSuite(c)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := hasher.Hash(in)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// HashToScalar performs a transformation to encode bytes as a Scalar object in the
+// appropiate group.
+func (c *Ciphersuite) HashToScalar([]byte) (*Scalar, error) {
+	return &Scalar{}, nil
+}
+
+// RandomScalar samples a random scalar value from the field of scalars defined by the
+// group order.
+// TODO: not constant time
+func (c *Ciphersuite) RandomScalar() (*Scalar, error) {
+	N := c.Order()
+	bitLen := N.X.BitLen()
+	byteLen := (bitLen + 7) >> 3
+	buf := make([]byte, byteLen)
+
+	// rejection sampling
+	for {
+		_, err := io.ReadFull(rand.Reader, buf)
+		if err != nil {
+			return nil, errors.New("scalar generation failed")
+		}
+		// Mask to account for field sizes that are not a whole number of bytes.
+		var mask = []byte{0xff, 0x1, 0x3, 0x7, 0xf, 0x1f, 0x3f, 0x7f}
+		buf[0] = buf[0] & mask[bitLen%8]
+
+		// Check if scalar is in the correct range.
+		if new(big.Int).SetBytes(buf).Cmp(N.X) >= 0 {
+			continue
+		}
+		break
+	}
+
+	x := new(big.Int).SetBytes(buf)
+	return &Scalar{c, x}, nil
+}
+
+// ScalarMultBase performs a scalar multiplication of the Generator with some scalar
+// input
+func (c *Ciphersuite) ScalarMultBase(s *Scalar) (*Point, error) {
+	gen := c.Generator()
+
+	return gen.ScalarMult(s)
+}
+
+// ByteLength returns the ByteLength of objects associated with the Ciphersuite.
+func (c *Ciphersuite) ByteLength() int {
+	return (c.curve.Params().BitSize + 7) / 8
+}
+
+// Point is a representation of a group element. It has two coordinates of
+// *big.Int type.
+type Point struct {
+	c *Ciphersuite
+	x *big.Int
+	y *big.Int
+}
+
+// Element is the interface that represents points in a given
+// Ciphersuite instantiation.
+type Element interface {
+	// Returns a bool indicating that the Element is valid
+	IsValid() bool
+
+	// Returns a bool indicating that the Element is the identity
+	IsIdentity() bool
+
+	// Performs a scalar multiplication of the Element with some scalar
+	// input
+	ScalarMult(*Scalar) (*Point, error)
```

Comment:
The name of the interface is `Element`, but here we return `Point`. Is there a reason for the discrepancy?

Reply:
Just changed it ;)